### PR TITLE
refactor userAllowedNavigation

### LINKED ISSUE #1560: Refactor permissions and userAllowedNavigation
**What needs to be done? Why does it need to be done?**
A clear and concise description of what task needs to be accomplished and why.

**Additional context**
Add any other context about the task here.

### PR DESCRIPTION
**Pull request by jovergaag**
## Description of the change

The `userAllowedNavigation`, specifically the `getAllowedNavigation` was becoming very complex and difficult to debug. I simplified the logic, wrote some helper functions and moved it all into the `UserStore`. A few new tests for the helper functions are included, and the old tests for `getAllowedNavigation` have been ported over.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Configuration change (adjusts configuration to achieve some end related to functionality, development, performance, or security)

## Related issues

> Closes #1560

### Development

These boxes should be checked by the submitter prior to merging:

- [x] Manual testing against realistic data has been performed locally
- [ ] E2E have been run for Lantern changes ([Steps in the Readme](https://github.com/Recidiviz/pulse-dashboard#running-e2e-tests))

### Code review

These boxes should be checked by reviewers prior to merging:

- [x] This pull request has a descriptive title and information useful to a reviewer
- [x] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [x] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

File: src/RootStore/UserStore.ts [hunk 1]
```diff
@@ -190,8 +195,14 @@ export default class UserStore {
...
     const routePermissions = entries(this.userAppMetadata?.routes);
+    const routes = routePermissions.map((r) => {
+      const [fullRoute, permission] = r;
+      const [view, page] = fullRoute.split("_");
+      const route = view === "operations" ? view : page;
+      return [route, permission];
+    });
     // @ts-ignore
-    return routePermissions as RoutePermission[];
+    return routes as RoutePermission[];
```
Comment (macfarlandian) [inline src/RootStore/UserStore.ts:205]: i think maybe `const routes: RoutePermission[] = whatever` would provide a little more type safety than making an assertion at the end?

**Review by macfarlandian — APPROVED** [review]

File: src/RootStore/UserStore.ts [hunk 2]
```diff
@@ -190,8 +195,14 @@ export default class UserStore {
   get routes(): RoutePermission[] {
     if (!this.userAppMetadata?.routes) return [];
     const routePermissions = entries(this.userAppMetadata?.routes);
+    const routes = routePermissions.map((r) => {
+      const [fullRoute, permission] = r;
```
Comment (Catacola) [inline src/RootStore/UserStore.ts:199]: If you'd like, you can do the array destructuring right in the function arguments and avoid needing to make an intermediate variable.

```suggestion
    const routes = routePermissions.map(([fullRoute, permission]) => {
```

File: src/RootStore/UserStore.ts [hunk 3]
```diff
@@ -240,20 +251,53 @@ export default class UserStore {
...
+    const permission = this.getRoutePermission(pageName);
+    return permission || !pagesWithRestrictions?.includes(pageName);
+  }
+
+  getRoutePermission(route: string): boolean {
+    const routePermission = this.routes.find((r) => r[0] === route);
+    // If the route does not exist in the RoutePermissions object, default to false;
+    if (!routePermission) return false;
+    const [, permission] = routePermission;
+    return permission;
```
Comment (Catacola) [inline src/RootStore/UserStore.ts:300]:
```suggestion
    return routePermission[1];
```

**Review by Catacola — APPROVED** [review]
Very nice! Thanks for writing descriptive and complete tests.

**Pull request merged**